feat(ensindexer): ENSDb Writer Worker

tk-o opened the pull request:
Store ENSNode metadata (ENSIndexer Public Config, Indexing Status) into ENSDb.

## Suggested review order

### ENSNode Schema

New schema was added for ENSNode metadata at `packages/ensnode-schema/src/schemas/ensnode-metadata.schema.ts`.

### ENSNode SDK

A simple HTTP client was built at `packages/ensnode-sdk/src/ensindexer/client.ts`. It's useful for sending HTTP requests over the internal network (akin to in-memory function calls):
- `/health` endpoint — required to respond with HTTP Status OK before any other requests can be made.
- `/api/config` endpoint — returns in-memory `ENSIndexerPublicConfig` object.
- `/api/indexing/status` endpoint — returns in-memory `IndexingStatusResponse` object.

### ENSIndexer

#### ENSDb module
Includes:
- `makeDrizzle` factory function known from ENSApi, which creates a Drizzle client instance.
  - `apps/ensindexer/src/lib/ensdb/drizzle.ts`
- `ENSNodeMetadata` type defining data model associated with `ensnode-metadata.schema.ts`
  - `apps/ensindexer/src/lib/ensdb/ensnode-metadata.ts`
- `EnsDbClient` class for performing ENSDb reads and writes.
  - `apps/ensindexer/src/lib/ensdb/ensdb-client.ts`

#### ENSDb Writer Worker

Initialized at ENSIndexer process startup:
- `apps/ensindexer/ponder/src/ensdb-writer-worker.ts`

Waits for ENSIndexer to become healthy and then:
1. Attempts to upsert `ENSIndexerPublicConfig` into ENSDb.
   - This may fail when ENSDb already had other record for ENSIndexerPublicConfig stored before, and the new record was not compatible with the stored one.
   - This is a single attempt and if it fails, the ENSIndexer application will be terminated with error.
   - Also, the ENSDb Version info is stored into ENSDb at the same time as ENSIndexerPublicConfig record.
2. Attempts to upsert `CrossChainIndexingStatusSnapshot` into ENSDb.
  - This may fail if the indexing status was not available yet, or the omnichain indexing status was "unstarted".
  - This is a recurring attempt; any failed attempt will log an error and there will be another attempt.

Related to #1252

tk-o committed:
feat(ensindexer): ENSDb Writer Worker

Store ENSNode metadata (ENSIndexer Public Config, Indexing Status) into ENSDb.

tk-o committed:
feat(ensapi): use ENSDb query and mutation clients

tk-o committed:
feat(ensapi): improve ENSDb Writer Worker

tk-o committed:
docs(changeset): Includes schema for `ENSNodeMetadata`.

tk-o committed:
docs(changeset): Introduces a worker that writes serialized representations of ENSIndexer Public Config and Indexing Status to ENSDb.

tk-o committed:
feat(ensapi): tune logs

shrugs commented on apps/ensindexer/ponder/src/ensdb-writer-worker.ts:
```diff
@@ -0,0 +1,142 @@
+    // Read stored config and in-memory config.
+    // Note: we wrap read operations in pRetry to ensure all of them are
+    // completed successfully.
+    const [storedConfig, inMemoryConfig] = await pRetry(() =>
```
technically it think this should be 

```
await Promise.all([
  pRetry(...),
  pRetry(...)
])
```

this way if one of the reads fails, the other isn't unnecessarily re-fetched

tk-o replied:
Awesome suggestion, thanks 🚀

shrugs commented on apps/ensindexer/ponder/src/ensdb-writer-worker.ts:
```diff
@@ -0,0 +1,142 @@
+    // if the stored one is available
+    if (storedConfig) {
+      try {
+        validateENSIndexerPublicConfigCompatibility(storedConfig, inMemoryConfig);
```
this is a good check but maybe not even necessary? because ponder already will reject re-indexing if the ponder config differs from this schema's previously used ponder config. obviously ponder config !== ensindexerpublicconfig but it's pretty close — can we think of a situation where that wouldn't be the case?

all that said there's no harm in leaving this check here, so that's fine

shrugs commented on apps/ensindexer/ponder/src/ensdb-writer-worker.ts:
```diff
@@ -0,0 +1,142 @@
+
+      // Check if Indexing Status is in expected status.
+      if (omnichainSnapshot.omnichainStatus === OmnichainIndexingStatusIds.Unstarted) {
+        throw new Error("Omnichain Status  must be different that 'Unstarted'.");
```
than* unstarted

shrugs commented on apps/ensindexer/ponder/src/ensdb-writer-worker.ts:
```diff
@@ -0,0 +1,142 @@
+      // Upsert ENSIndexerPublicConfig into ENSDb.
+      await ensDbMutation.upsertIndexingStatus(snapshot);
+
+      logger.info("Indexing Status successfully stored in ENSDb.");
```
- add a once-per-process log at info level for "Indexing Status successfully stored in ENSDb."
- make this a trace-level call with something like "Indexing Status updated in ENSDb."

tk-o replied:
In fact, there's no need to log the trace for the update event. The lack of error log implies the successful update.

shrugs commented on apps/ensindexer/ponder/src/ensdb-writer-worker.ts:
```diff
@@ -0,0 +1,142 @@
+
+// Run ENSDb Writer Worker in a non-blocking way to
+// allow database migrations to proceed in the background.
+setTimeout(ensDbWriterWorker, 0);
```
don't need to use settimeout, i woudl just do:

```
ensDbWriterWorker()
  .then() // logger.info ENSdb Writer Worker started
  .cach() // logger.error(error), process.exit
```

shrugs commented on apps/ensindexer/src/config/compatibility.ts:
```diff
@@ -0,0 +1,64 @@
+ * Validate if `configB` is compatible with `configA`, such that `configA` is
+ * a subset of `configB`.
+ *
+ * @throws error if 'indexedChainIds' were incompatible.
```
nit: simplify to `@throws error if configs are incompatible`

shrugs commented on apps/ensindexer/src/lib/ensindexer/client.ts:
```diff
@@ -0,0 +1,68 @@
+      logger.debug("Fetching ENSIndexer Health status: healthy");
+    } catch {
+      const errorMessage = "Health endpoint for ENSIndexer is not available yet.";
+      logger.error(errorMessage);
```
i don't think the client should log errors — the consumer should

shrugs commented on apps/ensindexer/src/lib/ensdb/drizzle.ts:
```diff
@@ -0,0 +1,41 @@
+// This file was copied 1-to-1 from ENSApi.
```
likewise for `logger.ts`

tk-o commented on apps/ensindexer/ponder/src/ensdb-writer-worker.ts:
```diff
@@ -0,0 +1,142 @@
+      // Upsert ENSIndexerPublicConfig into ENSDb.
+      // Note: we wrap write operation in pRetry to ensure it can complete
+      // successfully, as there will be no other attempt.
+      await pRetry(() => ensDbMutation.upsertEnsIndexerPublicConfig(inMemoryConfig));
```
I didn't feel 100% ok with this pRetry, perhaps we should the process fail if there's any (very unlikely) issue with the write 👍

tk-o committed:
apply pr feedback

Remove logs from ENSIndexerClient class

tk-o commented on apps/ensindexer/src/lib/ensdb/ensdb-mutation.ts:
```diff
@@ -0,0 +1,75 @@
+import { type EnsNodeMetadata, EnsNodeMetadataKeys } from "./ensnode-metadata";
+
+/**
+ * ENSDb Mutation
```
Will do 👍

tk-o committed:
apply pr feedback

consolidate ensdb modules; replace pino logger with native console logger; simplify ensdb worker logic

lightwalker-eth commented on apps/ensindexer/src/lib/ensindexer/client.ts:
```diff
@@ -0,0 +1,51 @@
+  SerializedIndexingStatusResponse,
+} from "@ensnode/ensnode-sdk";
+
+export class EnsIndexerClient {
```
I imagine this should move into the `ensnode-sdk` package?

It's true that we would only use it ourselves within ENSIndexer, but in theory we should put the clients for each of the ENSNode apps into `ensnode-sdk`. At least with my current mental model.

lightwalker-eth commented on packages/ensnode-schema/src/ponder.schema.ts:
```diff
@@ -2,6 +2,7 @@
  * Merge the various sub-schemas into a single ponder (drizzle) schema.
```
In my mental model:

1. There should be absolutely no references to "ponder" at all within this package.
2. All ponder dependencies in this package should be transferred into `ensindexer`. The only thing in our whole monorepo that should have any dependency on Ponder should be the `ensindexer` app. No other app or package should have any Ponder deps.
3. This package should be renamed from `ensnode-schema` to `ensdb-sdk` or something like that. It would not only hold the Drizzle schemas, but it would also be the home for the clients we'll be working to build up that would contain the logic for db read and write operations used across the ENSNode monorepo. Not trying to suggest we completely transfer all db read and write operations into this package within this PR. That's too much for this PR. But sharing the idea here to help share an idea on overall future direction.

What do you think? Appreciate your advice.

tk-o replied:
Removing `ponder` dependency from `ensnode-schema` package will possibly require maintaining our own implementation of this directory: https://github.com/ponder-sh/ponder/tree/ponder%400.13.14/packages/core/src/drizzle

I don't think it's practical at the moment, given no plans for ENSNode to migrate away from `ponder`.

lightwalker-eth commented on pnpm-workspace.yaml:
```diff
@@ -8,6 +8,7 @@ catalog:
   "@astrojs/tailwind": ^6.0.2
   "@namehash/namekit-react": 0.12.0
+  "@ponder/client": 0.13.14
   "@ponder/utils": 0.2.14
```
Please see other comments suggesting all ponder deps would exclusively live in `ensindexer`. What do you think?

tk-o replied:
If we want to remove these dependencies, we need to implement the functionality provided by these packages on our own:
- `@ponder/client`
  - `setDatabaseSchema`
- `@ponder/utils`
  - `mergeAbis`
  - `replaceBigInts`

lightwalker-eth commented on apps/ensindexer/src/lib/ensindexer/client.ts:
```diff
@@ -0,0 +1,51 @@
+  constructor(private ensIndexerUrl: URL) {}
+
+  /**
+   * Wait for ENSIndexer to become healthy.
```
We should document here:

1. What does it represent for ENSIndexer to be healthy? What does this guarantee about state?
2. What does it represent for ENSIndexer not to be healthy? What should someone avoid doing if ENSIndexer isn't healthy yet?

lightwalker-eth commented on apps/ensindexer/src/lib/ensindexer/client.ts:
```diff
@@ -0,0 +1,51 @@
+   *
+   * @throws when the Health endpoint didn't return HTTP Status OK.
+   */
+  public async health(): Promise<void> {
```
Shouldn't there be a distinct representation of each of the following possible results?
1. explicitly received a response, the HTTP response code wasn't 200 OK though.
2. never received any response

tk-o replied:
Implemented `EnsIndexerHealthCheckResults` to handle all possible response result scenarios.

lightwalker-eth commented on apps/ensindexer/src/lib/ensindexer/client.ts:
```diff
@@ -0,0 +1,51 @@
+  /**
+   * Fetch Indexing Status
+   *
+   * @returns Indexing Status when it's available and with valid.
```
Suggest to refine this sentence.

lightwalker-eth commented on apps/ensindexer/src/lib/ensindexer/local-client.ts:
```diff
@@ -0,0 +1,16 @@
+/**
+ * Wait for ENSIndexer to become healthy.
+ *
+ * The global promise that will only resolve after the ENSIndexer has become healthy.
```
I'm not familiar with `pRetry`. If the 5 retries all fail, is it confirmed then it causes ENSIndexer to terminate itself? Appreciate your advice.

lightwalker-eth commented on apps/ensindexer/src/lib/ensindexer/local-client.ts:
```diff
@@ -0,0 +1,16 @@
+ * The global promise that will only resolve after the ENSIndexer has become healthy.
+ */
+export const waitForEnsIndexerToBecomeHealthy = pRetry(async () => ensIndexerClient.health(), {
+  retries: 5,
```
How do you suggest we handle the ideas discussed in our team call earlier today about ENSIndexer potentially needing to wait up to an hour for ENSRainbow to become both healthy and ready?

tk-o replied:
This is idea for a separate PR, we need to update the way we use `EnsRainbowApiClient` inside `ensindexer` app, such that, before calling a method on an `ensRainbowApiClient` instance, we await for a health check promise to resolve. It should be similar to how we use `waitForEnsIndexerToBecomeHealthy` promise in the `EnsIndexerClient` use case.

lightwalker-eth commented on apps/ensindexer/src/lib/ensdb/ensdb-client.ts:
```diff
@@ -0,0 +1,161 @@
+  });
+
+  /**
+   * Upsert ENSIndexer Public Config
```
```suggestion
   * Get ENSIndexer Public Config
```

lightwalker-eth commented on apps/ensindexer/src/lib/ensdb/ensdb-client.ts:
```diff
@@ -0,0 +1,161 @@
+  /**
+   * Upsert ENSIndexer Public Config
+   *
+   * @returns updated record in ENSDb.
```
```suggestion
   * @returns ENSIndexer public config in ENSDb.
```

lightwalker-eth commented on apps/ensindexer/src/lib/ensdb/ensdb-client.ts:
```diff
@@ -0,0 +1,161 @@
+   * Upsert ENSIndexer Public Config
+   *
+   * @returns updated record in ENSDb.
+   * @throws when upsert operation failed.
```
```suggestion
   * @throws when read operation failed.
```

lightwalker-eth commented on apps/ensindexer/src/lib/ensdb/ensdb-client.ts:
```diff
@@ -0,0 +1,161 @@
+   * @returns updated record in ENSDb.
+   * @throws when upsert operation failed.
+   */
+  async getEnsIndexerPublicConfig(): Promise<SerializedENSIndexerPublicConfig | undefined> {
```
In my mind, the serialization / deserialization responsibilities should live within this class. Therefore all params passed into all the functions in this file or returned by the functions in this file should be deserialized business-logic-level object types.

This comment applies across the whole file.

shrugs reviewed (approved):
LGTM!

lightwalker-eth commented on apps/ensindexer/ponder/src/ensdb-writer-worker.ts:
```diff
@@ -0,0 +1,131 @@
+ *    into ENSDb.
+ */
+async function ensDbWriterWorker() {
+  console.log("ENSDb Writer Worker: waiting for ENSIndexer to become healthy.");
```
What do you think about potentially having Ponder wait up to an hour to have us pass it a Ponder config while we wait for ENSRainbow to become healthy / ready such that ENSIndexer can complete building its config?

lightwalker-eth commented on apps/ensindexer/ponder/src/ensdb-writer-worker.ts:
```diff
@@ -0,0 +1,131 @@
+  console.log("ENSDb Writer Worker: ENSIndexer is healthy, starting tasks.");
+
+  // 1. Create ENSDb Client
+  const ensDbClient = new EnsDbClient();
```
Hmm. I think we need to give more of a detailed focus on the case of a completely fresh `DATABASE_SCHEMA` and how this is initialized by Ponder.

We have the existing code path:
1. ENSIndexer fetches ENSRainbow config from ENSRainbow
2. ENSIndexer can now build its own config
3. ENSIndexer uses its own config to build a Ponder config
4. Ponder uses the Ponder config to initialize `DATABASE_SCHEMA`.

Right? Don't we need to wait until step 4 for the tables in `DATABASE_SCHEMA` to be initialized?

Assuming so, this waiting potentially needs to grow a lot if we add logic to wait up to an hour for step 1 above to be complete, waiting for ENSRainbow to become both healthy and ready.

And during all of this time, what is ENSApi supposed to do? From its perspective, it's been configured to read from `DATABASE_SCHEMA` but ENSIndexer might wait over an hour to create it.

Suggest to take a step back and carefully consider all cases and situations, then to document a plan and share it with the team for review 👍

tk-o replied:
I like the idea for ENSApi to use specific `DATABASE_SCHEMA`. Suggest sticking to that.

ENSApi has to wait for ENSDb to include `EnsNodeMetadata`. Otherwise, there's literally no use case ENSApi could serve. All ENSApi routes require indexing status information, which come from ENSDb.

ENSDb won't have the `EnsNodeMetadata` information until ENSIndexer stores it.

ENSIndexer won't store any `EnsNodeMetadata` information until ENSRainbow is healthy, and ENSIndexer is healthy.

I think we should create a "logical healthcheck" for ENSDb, such that clients connecting to ENSDb could know if the service is ready to be used. For example, we can consider this "logical healthcheck" to ensure that ENSDb includes `EnsNodeMetadata` for both keys: `EnsNodeMetadataKeys.EnsIndexerPublicConfig`, `EnsNodeMetadataKeys.IndexingStatus`.

lightwalker-eth reviewed (commented):
@tk-o Thanks for this. Reviewed and shared feedback 👍

tk-o commented on pnpm-workspace.yaml:
```diff
@@ -18,6 +19,8 @@ catalog:
+  p-retry: ^7.1.1
   pg-connection-string: ^2.9.1
   pino: 10.1.0
   ponder: 0.13.16
```
Right now, we use `ponder` in:
- `packages/ensnode-schema`
  - Removing `ponder` dependency here requires further consideration, as explained [here](https://github.com/namehash/ensnode/pull/1406#discussion_r2630426712) and [there](https://github.com/namehash/ensnode/pull/1406#discussion_r2630342844).
- `packages/ponder-metadata`
  - It should be easy to remove `ponder` as dependency here 👍 Will create a follow-up PR including that change and that also resolves https://github.com/namehash/ensnode/issues/908

tk-o commented on apps/ensindexer/src/lib/ensindexer/client.ts:
```diff
@@ -0,0 +1,51 @@
+      new URL("/api/config", this.ensIndexerUrl),
+    ).then((response) => response.json());
+
+    return ensIndexerPublicConfigSerialized as SerializedENSIndexerPublicConfig;
```
>Perhaps the idea in this comment is too big for this PR and should be moved into a separate follow-up issue?
>
>I'm curious why we are returning serialized types from our clients. In my mental model, all of our clients should also take on the responsibility of deserialization and returning business-logic-level objects.

I'll follow up on this request in another PR for which the scope was mentioned [here](https://github.com/namehash/ensnode/pull/1406#discussion_r2630311539). TLDR: we need to split request and response types between `EnsIndexerClient` and `EnsApiClient` (which is called `ENSNodeClient` as of now).

> I'm also curious about the use of https://hono.dev/docs/guides/rpc (hono client, internally within our own client implementations so that hono isn't directly exposed / exported by our packages) and how that relates to the work in progress by @stevedylandev in this PR: https://github.com/namehash/ensnode/pull/1338

@lightwalker-eth could you please expand on specific request for me here? I don't understand the expectation here.

tk-o commented on apps/ensindexer/src/lib/ensdb/ensdb-client.ts:
```diff
@@ -0,0 +1,161 @@
+   * @returns updated record in ENSDb.
+   * @throws when upsert operation failed.
+   */
+  async upsertIndexingStatus(
```
Answered [here](https://github.com/namehash/ensnode/pull/1406#discussion_r2630593201).

tk-o replied:
I updated all "upsert" methods to have `Promise<void>` return type.

tk-o committed:
Update EnsDbClient and EnsIndexerClient with serialization/deserialization logic

tk-o commented on apps/ensindexer/src/lib/ensdb/ensnode-metadata.ts:
```diff
@@ -0,0 +1,31 @@
+  SerializedCrossChainIndexingStatusSnapshot,
+  SerializedENSIndexerPublicConfig,
+} from "@ensnode/ensnode-sdk";
+
```
Will do that in a follow up PR.

tk-o commented on apps/ensindexer/src/lib/ensdb/drizzle.ts:
```diff
@@ -0,0 +1,25 @@
+// This file was copied 1-to-1 from ENSApi.
+// TODO: deduplicate with apps/ensapi/src/lib/handlers/drizzle.ts when ensnode nodejs internal package is created
```
Using `setDatabaseSchema` is more like a fix for `drizzle` object instance to use the specified `schemaName` for queries.

tk-o committed:
Move `EnsIndexerClient` into ENSNode SDK

tk-o committed:
Fix import paths to avoid circular dependencies

tk-o commented on apps/ensindexer/src/lib/ensdb/ensdb-client.ts:
```diff
@@ -0,0 +1,161 @@
+   */
+  async upsertEnsIndexerPublicConfig(
+    ensIndexerPublicConfig: SerializedENSIndexerPublicConfig,
+  ): Promise<SerializedENSIndexerPublicConfig> {
```
I updated all "upsert" methods to have `Promise<void>` return type.

tk-o commented on packages/ensnode-sdk/src/ensindexer/config/compatibility.ts:
```diff
@@ -0,0 +1,61 @@
+import type { SerializedENSIndexerPublicConfig } from "@ensnode/ensnode-sdk";
+
```
Nice idea, will have it moved to be co-located with `EnsIndexerClient` file.

tk-o committed:
Include ENSDb Version in ENSNodeMetadata table

lightwalker-eth commented on packages/ensnode-schema/src/schemas/ensnode-metadata.schema.ts:
```diff
@@ -0,0 +1,19 @@
+
+import { onchainTable } from "ponder";
+
+export const ensNodeMetadata = onchainTable("ensnode_metadata", (t) => ({
```
@tk-o Thanks for this detailed investigation and reply 👍 

I've transferred these ideas into https://github.com/namehash/ensnode/issues/1234#issuecomment-3675295303 for future follow up and will mark this conversation thread as resolved for the scope of this PR.

lightwalker-eth commented on apps/ensindexer/src/lib/ensindexer/local-client.ts:
```diff
@@ -2,7 +2,13 @@ import config from "@/config";
+/**
+ * How many times retries should be attempted before
+ * {@link waitForEnsIndexerToBecomeHealthy} becomes a rejected promise.
+ */
```
```suggestion
/**
 * Upon the startup of ENSIndexer, defines the max number failures ENSIndexer is allowed to experience
 * when attempting to get a successful response from its own `/health` endpoint. If these max retries are exceeded
 * {@link waitForEnsIndexerToBecomeHealthy} becomes a rejected promise
 * which ultimately results in ENSIndexer crashing itself.
 */
```

Is that fair?

lightwalker-eth commented on apps/ensindexer/src/lib/ensindexer/local-client.ts:
```diff
@@ -11,6 +17,15 @@ export const ensIndexerClient = new EnsIndexerClient(config.ensIndexerUrl);
+    const response = await ensIndexerClient.health();
+
+    if (response !== EnsIndexerHealthCheckResults.Ok) {
+      throw new Error("ENSIndexer is not healthy yet");
```
Would it be helpful to include a log message in here too? Please ignore if not relevant.

lightwalker-eth commented on apps/ensindexer/ponder/src/ensdb-writer-worker.ts:
```diff
@@ -0,0 +1,134 @@
+ * This file manages syncing ENSNode metadata:
+ * - ENSIndexer Public Config
+ * - Indexing Status
+ * into the ENSDb.
```
```suggestion
 * This file manages upserting ENSNodeMetadata into ENSDb.
```

lightwalker-eth commented on apps/ensindexer/ponder/src/ensdb-writer-worker.ts:
```diff
@@ -0,0 +1,134 @@
+ *    upsert serialized representation of {@link CrossChainIndexingStatusSnapshot}
+ *    into ENSDb.
+ */
+async function ensDbWriterWorker() {
```
```suggestion
async function ensDbMetadataUpsertWorker() {
```

Is that fair? Seems much more accurate.

Assuming so, please rename all the related things / comments / files / etc.

lightwalker-eth commented on apps/ensindexer/ponder/src/ensdb-writer-worker.ts:
```diff
@@ -0,0 +1,134 @@
+  OmnichainIndexingStatusIds,
+  validateENSIndexerPublicConfigCompatibility,
+} from "@ensnode/ensnode-sdk";
+
```
Why does this file live under the `/ponder` directory?

lightwalker-eth commented on apps/ensindexer/ponder/src/ensdb-writer-worker.ts:
```diff
@@ -0,0 +1,134 @@
+
+      // Check if Indexing Status is available.
+      if (inMemoryIndexingStatus.responseCode !== IndexingStatusResponseCodes.Ok) {
+        throw new Error("Indexing Status must be available.");
```
We need to write better. This should explain why.

lightwalker-eth commented on apps/ensindexer/ponder/src/ensdb-writer-worker.ts:
```diff
@@ -0,0 +1,134 @@
+  /**
+   * Handle ENSIndexerPublicConfig and ENSDb Version Records
+   */
+  const handleEnsIndexerPublicConfigAndEnsDbVersionRecords = async () => {
```
I note how this function doesn't initialize the metadata field for indexing status.

As a result, ENSDb clients have additional complexity in that they need to prepare for a situation where there may be a long duration of time where they can read 2 of the 3 metadata fields, but not the indexing status field.

What do you think about this? Appreciate your advice. Should we update this function so that it waits until all 3 metadata fields can be initialized at the same time, and then initializes them all together?

lightwalker-eth commented on apps/ensindexer/src/lib/ensindexer/local-client.ts:
```diff
@@ -0,0 +1,31 @@
+ *
+ * The global promise that will only resolve after the ENSIndexer has become healthy.
+ */
+export const waitForEnsIndexerToBecomeHealthy = pRetry(
```
Is there a special reason to put this function in this file? In my mind it seems it would be better moved into `ensdb-writer-worker.ts`.

lightwalker-eth reviewed (commented):
@tk-o Thanks for updates. Reviewed and shared feedback.
